client/signup/steps/user/index.jsx

```diff
@@ -115,6 +115,8 @@ export class UserStep extends Component {
 			locale: this.props.locale,
 			oauth2ClientId: this.props.oauth2Client && this.props.oauth2Client.id,
 			wccomFrom: this.props.wccomFrom,
+			isWhiteLogin: this.props.isReskinned,
+			signupUrl: window.location.pathname + window.location.search,
```

Comment:
I suggest to add optional chaining to ensure the code from breaking down. 
```suggestion
			signupUrl: window.location?.pathname + window.location?.search,
```

Reply:
Interesting! Under what conditions would the code break down here you think? I'm thinking something like that would happen either in SSR (where `window` may be undefined?) or tests (probably mistakenly)? 🤔

Reply:
Well, the window API is in a sense a form of user input so we have to be careful trusting it. And you're right that it might break in SSR and tests.

Reply:
> Well, the window API is in a sense a form of user input so we have to be careful trusting it.

Hmm `location` is a read-only property, so an attempt to delete it from context will result in a no-op. So unless we purposefully shadow `window` (in some closure or part of the build), then the use of it feels pretty safe to me. As for whether we trust that a certain browser will provide an implementation for it, I believe our browser list is also safe. Happy to learn otherwise though. Do you have a specific example in mind?

> And you're right that it might break in SSR and tests.

Not so sure I am right. 😅 The suggested change obviously won't help in case the Window context itself is undefined. I am not sure we should be concerned with either of these at this point in the app.

p.s. We use `window.location.*` quite heavily in Calypso (including in this file), so it might be worth moving the conversation to an issue. From my understanding, optional chaining on `location` feels unnecessary here. Happy/curious to learn otherwise though. :-)

Reply:
If you want to avoid browser APIs, you can get the route info from the `getCurrentRoute` and `getCurrentQueryArguments` Redux selectors. That will work even on the server. But Signup doesn't use SSR so it doesn't matter that much.

`window.location`, if you choose to keep it, is a browser API that's been around forever. I don't see a reason to guard calls with `window.location?.search`.

Reply:
Thanks for the reply @jsnajdr 

> If you want to avoid browser APIs, you can get the route info from the `getCurrentRoute` and `getCurrentQueryArguments` Redux selectors. That will work even on the server. But Signup doesn't use SSR so it doesn't matter that much.

Interesting (and weirdly hard) choice to make. Not sure I have a preference. `window.location` feels like such a fundamental concept for simple uses like above. I'd also be happy with using an abstraction instead. Do you have a preference?

Reply:
My preference would be to create a new selector, `createCurrentRouteURL`, that constructs the entire `/path?query` string, and use that.

Imagine rendering the React component in an Express.js server, producing a static HTML response. There, you know the request URL: it's in the HTTP request and in the server's `Request` object. And you can use it to render the right `signupUrl` link. There's no `window` and no `window.location` though. The route information is available from a different source. Here, the Redux selector (or any other abstraction mechanism, e.g., some custom `useCurrentRoute()` hook) provides a unified way how to read the current route.

Reply:
> My preference would be to create a new selector, `createCurrentRouteURL`, that constructs the entire `/path?query` string, and use that.

Makes sense. Constructing `/path?query` away from the API will inevitably lead to a separate selector. There's logic to bring these parts together from state. Not sure about the name `createCurrentRouteURL `, but that's a detail we can work out later. We have a few instances of `window.location.pathname + window.location.search `, so it might be a step in the right direction. 

> Imagine rendering the React component in an Express.js server, producing a static HTML response. There, you know the request URL: it's in the HTTP request and in the server's `Request` object. And you can use it to render the right `signupUrl` link. There's no `window` and no `window.location` though. The route information is available from a different source. Here, the Redux selector (or any other abstraction mechanism, e.g., some custom `useCurrentRoute()` hook) provides a unified way how to read the current route.

Sounds like an ideal scenario. At the moment, take out `window` and I sense 99% of Calypso would fall apart (including this component). 😃 Small steps making a whole I guess.

I'll resolve this thread and pick things up in a separate PR. We'll stick to the current approach for now.

Reply:
Thanks for sparking an interesting convo here @StefanNieuwenhuis. 💯 

Marked as resolved for now. There's no need to guard `location` if we have a browser/Window context. It's read-only and around since the age of dinosaurs, so should be pretty safe. I'll try and prioritize the selector that @jsnajdr mentioned - it should be fun.